include/core/mir/geometry/point_generic.h

```diff
@@ -0,0 +1,94 @@
+ */
+
+#ifndef MIR_GEOMETRY_POINT_GENERIC_H_
+#define MIR_GEOMETRY_POINT_GENERIC_H_
+
+#include "dimensions_generic.h"
+
+namespace mir
+{
+namespace geometry
+{
+namespace detail
+{
+struct PointBase{}; ///< Used for determining if a type is a point
+}
+namespace generic
+{
+template<template<typename> typename T>
+struct Point : detail::PointBase
+{
+    template<typename Tag>
+    using WrapperType = T<Tag>;
+
+    constexpr Point() = default;
+    constexpr Point(Point const&) = default;
+    Point& operator=(Point const&) = default;
+
+    template<typename XType, typename YType>
+    constexpr Point(XType&& x, YType&& y) : x(x), y(y) {}
+
+    T<XTag> x;
+    T<YTag> y;
+};
+
+template<typename P, typename std::enable_if<std::is_base_of<detail::PointBase, P>::value, bool>::type = true>
+inline constexpr bool operator == (P const& lhs, P const& rhs)
+{
+    return lhs.x == rhs.x && lhs.y == rhs.y;
+}
+
+template<typename P, typename std::enable_if<std::is_base_of<detail::PointBase, P>::value, bool>::type = true>
+inline constexpr bool operator != (P const& lhs, P const& rhs)
+{
+    return lhs.x != rhs.x || lhs.y != rhs.y;
+}
+
+template<typename P, typename std::enable_if<std::is_base_of<detail::PointBase, P>::value, bool>::type = true>
+inline constexpr P operator+(P lhs, typename P::template WrapperType<DeltaXTag> rhs) { return{lhs.x + rhs, lhs.y}; }
+template<typename P, typename std::enable_if<std::is_base_of<detail::PointBase, P>::value, bool>::type = true>
+inline constexpr P operator+(P lhs, typename P::template WrapperType<DeltaYTag> rhs) { return{lhs.x, lhs.y + rhs}; }
+
+template<typename P, typename std::enable_if<std::is_base_of<detail::PointBase, P>::value, bool>::type = true>
+inline constexpr P operator-(P lhs, typename P::template WrapperType<DeltaXTag> rhs) { return{lhs.x - rhs, lhs.y}; }
+template<typename P, typename std::enable_if<std::is_base_of<detail::PointBase, P>::value, bool>::type = true>
+inline constexpr P operator-(P lhs, typename P::template WrapperType<DeltaYTag> rhs) { return{lhs.x, lhs.y - rhs}; }
+
+template<typename P, typename std::enable_if<std::is_base_of<detail::PointBase, P>::value, bool>::type = true>
+inline P& operator+=(P& lhs, typename P::template WrapperType<DeltaXTag> rhs) { lhs.x += rhs; return lhs; }
+template<typename P, typename std::enable_if<std::is_base_of<detail::PointBase, P>::value, bool>::type = true>
+inline P& operator+=(P& lhs, typename P::template WrapperType<DeltaYTag> rhs) { lhs.y += rhs; return lhs; }
+
+template<typename P, typename std::enable_if<std::is_base_of<detail::PointBase, P>::value, bool>::type = true>
+inline P& operator-=(P& lhs, typename P::template WrapperType<DeltaXTag> rhs) { lhs.x -= rhs; return lhs; }
+template<typename P, typename std::enable_if<std::is_base_of<detail::PointBase, P>::value, bool>::type = true>
+inline P& operator-=(P& lhs, typename P::template WrapperType<DeltaYTag> rhs) { lhs.y -= rhs; return lhs; }
```

Comment:
This use of `typename P::template WrapperType<Foo>` is an "enable_if" alternative?

Reply:
I guess?? I'm not completely sure how this would be replaced by `enable_if`, but there's probably a way.

Reply:
NM On looking closer I see what's going on.